pkg/ebpf/c/tracee.bpf.c

```diff
@@ -6289,3 +6313,836 @@ int tc_ingress(struct __sk_buff *skb)
+//
+
+SEC("kprobe/__cgroup_bpf_run_filter_skb")
+int BPF_KPROBE(cgroup_bpf_run_filter_skb)
+{
+    // runs BEFORE the CGROUP/SKB eBPF program
+
+    event_data_t data = {};
+
+    if (!init_event_data(&data, ctx))
+        return 0;
+
+    // Both ingress & egress (mostly ingress) might run from a kthread and still
+    // need processing. Instead of "should_trace", check if current socket is
+    // a socket being traced or not.
+
+    struct sock *sk = (void *) PT_REGS_PARM1(ctx);
+    struct sk_buff *skb = (void *) PT_REGS_PARM2(ctx);
+    int type = PT_REGS_PARM3(ctx);
+
+    // obtain socket inode
+    u64 inode = BPF_READ(sk, sk_socket, file, f_inode, i_ino);
+    if (inode == 0)
+        return 0;
+
+    switch (type) {
+        case BPF_CGROUP_INET_INGRESS:
+        case BPF_CGROUP_INET_EGRESS:
+            break;
+        default:
+            return 0; // wrong attachment type, return fast
+    }
+
+    // save args for kretprobe
+    struct entry entry = {0};
+    entry.args[0] = PT_REGS_PARM1(ctx); // struct sock *sk
+    entry.args[1] = PT_REGS_PARM2(ctx); // struct sk_buff *skb
+
+    // prepare for kretprobe using entrymap
+    u32 pid = data.context.task.host_pid;
+    bpf_map_update_elem(&entrymap, &pid, &entry, BPF_ANY);
+
+    // NOTE: The net_task_context is the event_context's "task_context" when the
+    //       socket was created. This could take tracee to situations which the
+    //       event being submited by the cgroup skb ingress/egress eBPF program
+    //       has outdated information (host_ppid, ppid, comm, cgroup_id, uid,
+    //       mnt_id, pid_id, uts_name). The only possible way to solve this is
+    //       to hook to (*file)->fops() functions so, each time a socket is
+    //       poll'ed(), written(), read(), sendmsg'ed(), the task reading or
+    //       writing to the socket will have its task_info updated with correct
+    //       "should_trace" metadata (and then its net_task_context counter part
+    //       could also be updated).
+    //
+    //       Hooking into a socket file's file operations handlers might be
+    //       expensive in terms of performance, and it is fair to assume long
+    //       lasting sockets would be traced with old task context (from the
+    //       time the socket was created). So, for now, this is A DECISION and
+    //       not a lack of support.
+    //
+    //       When a task has an opened socket (or "long lasting" socket), and it
+    //       is migrated to a container that is currently being traced, for
+    //       example, whenever its "task_info" is updated, the should_trace
+    //       logic will allow the cgroup skb egress (not ingress, since it runs
+    //       in a kernel thread context) ebpf programs to start tracing those
+    //       sockets. This is the same technique used for existing sockets of
+    //       tasks that does not have their "task_info" updated.
```

Comment:
I'll rephrase the comment (to remove the emphasis on "the only way") and I have opened the issue: https://github.com/aquasecurity/tracee/issues/2221 to track this need (and further discussions).

Reply:
Rephrased to:
```
    // NOTE: The net_task_context is the event_context's "task_context" when the
    //       socket was created. This could take tracee to situations which the
    //       event being submited by the cgroup skb ingress/egress eBPF program
    //       has outdated information (host_ppid, ppid, comm, cgroup_id, uid,
    //       mnt_id, pid_id, uts_name). One possible way to solve this is
    //       to hook to (*file)->fops() functions so, each time a socket is
    //       poll'ed(), written(), read(), sendmsg'ed(), the task reading or
    //       writing to the socket will have its task_info updated with correct
    //       "should_trace" metadata (and then its net_task_context counter part
    //       could also be updated).
    //
    //       Hooking into a socket file's file operations handlers might be
    //       expensive in terms of performance, and it is fair to assume long
    //       lasting sockets would be traced with old task context (from the
    //       time the socket was created). So, for now, this is A DECISION and
    //       not a lack of support.
    //
    //       TODO: https://github.com/aquasecurity/tracee/issues/2221
    //
    //       Current mitigation: When a task has an opened socket (or "long
    //       lasting" socket), and it is migrated to a container that is
    //       currently being traced, for example, whenever its "task_info" is
    //       updated, the should_trace logic will allow the cgroup skb egress
    //       (not ingress, since it runs in a kernel thread context) ebpf
    //       programs to start tracing those sockets. This is the same technique
    //       used for existing sockets of tasks that does not have their
    //       "task_info" updated.
```